### ISSUE TITLE
Undefined JNA version in oshi-core

### ISSUE THREAD

**Issue by INPUTsys-Chris** [oshi/oshi#2300]
Oshi-core defines JNA version as: ${jna.version}

Since this checkin the value is only present in parent pom:
https://github.com/oshi/oshi/commit/31d8bbee4f6a0c82d28938915ad870bfc6ecef2c#diff-46a0651fc0e86287622e5439d7647fbf5eed17b90e43b3168622f82273a1a4a8

This leads to unspecified behaviour. Maven seems to pickup older JNA versions if another dependency with older JNA is present.

MVNRepository also can't determine which version should be used:
https://mvnrepository.com/artifact/com.github.oshi/oshi-core/6.4.0

**Comment by hazendaz** [comment]
Not how maven works.  The correct version is properly resolved in this project.  When a consumer uses elsewhere, the maven reactor determines the version based on many variables including use of dependency management of all used artifacts, order, etc.  If you find jna resolves wrong its due to that not usage of a property value.  In that case adjust your order or add dependency management to force the version to be locked in.

Not sure what you mean on reference points, jna is resolved there just fine as that straight reads the pom.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: INPUTsys ***@***.***>
Sent: Tuesday, January 31, 2023 4:16:09 AM
To: oshi/oshi ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [oshi/oshi] Undefined JNA version in oshi-core (Issue #2300)


Oshi-core defines JNA version as: ${jna.version}

Since this checkin the value is only present in parent pom:
31d8bbe#diff-46a0651fc0e86287622e5439d7647fbf5eed17b90e43b3168622f82273a1a4a8<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Foshi%2Foshi%2Fcommit%2F31d8bbee4f6a0c82d28938915ad870bfc6ecef2c%23diff-46a0651fc0e86287622e5439d7647fbf5eed17b90e43b3168622f82273a1a4a8&data=05%7C01%7C%7C29dc953d4de8453e683608db036bca62%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638107533831799899%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pVc3NuCGG8b965TvJJl3uIiR7xrAfDdTMvcw3zMck5o%3D&reserved=0>

This leads to unspecified behaviour. Maven seems to pickup older JNA versions if another dependency with older JNA is present.

MVNRepository also can't determine which version should be used:
https://mvnrepository.com/artifact/com.github.oshi/oshi-core/6.4.0<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fmvnrepository.com%2Fartifact%2Fcom.github.oshi%2Foshi-core%2F6.4.0&data=05%7C01%7C%7C29dc953d4de8453e683608db036bca62%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638107533831799899%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=TilmfXgbj%2BFqJP8EzPuv%2BQaEXFpAtAO%2FBLTmt74Q6H8%3D&reserved=0>

**Comment by INPUTsys-Chris** [comment]
Order doesn't matter. I guess maven prefers the version included in the dependencies instead of looking up oshi-parent.

Here is a sample project:
https://github.com/vaadin/flow/issues/15782

The pity is that maven omits the older Oshi 6.1.6 in favor of 6.4 but then uses the older dependency on JNA 5.11 instead of 5.12.1.

![grafik](https://user-images.githubusercontent.com/56319603/215758918-716ee881-d431-4822-9051-c0e953d3b82f.png)

Maybe I'm not smart enough for that kind of maven problems but I'll manage JNA dependency for myself in the future.

Thank you

**Comment by dbwiddis** [comment]
Hey @INPUTsys-Chris thanks for the report.  I'm happy to try to fix this but I still don't quite understand the problem, and I'm not quite sure how to reproduce it: you reference a "sample project" but that's just an issue on another project.  I downloaded and extracted the zip file linked in that issue but don't see either OSHI or JNA mentioned in it either.  Any chance you can throw together a small github project with a pom demonstrating the bug?

The versions for SLF4J and JUnit are similarly defined in the parent and they seem to be importing just fine and showing up on mvnrepository.com. 

What is different between these is that OSHI's versions are defined as part of `<dependencyManagement>` but the versions are not repeated in the `<dependencies>` section.  

This would seem to imply getting rid of the `<dependencyManagement>` declaration and defining the normal dependencies (but leaving the property in the parent) may be the correct solution, but without a reproducible project to test I'm not sure of that either.

Which brings up this point:

> When a consumer uses elsewhere, the maven reactor determines the version based on many variables including use of dependency management of all used artifacts, order, etc. 

Has something changed recently with this, @hazendaz ?  I'm pretty sure I would have noticed that JNA property placeholder some time ago, and I'm wondering if there's been a recent change in behavior in dependency processing.

**Comment by dbwiddis** [comment]
This Stack Overflow Question seems similar: https://stackoverflow.com/questions/28312975/maven-dependencymanagement-version-ignored-in-transitive-dependencies

**Comment by dbwiddis** [comment]
And, linked from the SO question, here's the Maven bug:  https://issues.apache.org/jira/browse/MNG-5761

I think the workaround (until maven 4) is to add an explicit version (even if it's the parent pom property) instead of relying on the dependency management, which is apparently useless in maven 3.x.

**Comment by dbwiddis** [comment]
Another solution for you to work around the maven bug, @INPUTsys-Chris : inherit OSHI's dependency management: https://stackoverflow.com/questions/45041888/how-can-i-depend-on-a-library-with-transitive-dependencies-which-are-adjusted-by

**Comment by dbwiddis** [comment]
Another related maven bug, probably the same root cause.  This bug has existed since 2015, someone has fixed it on their local github branch, the line of code is known, amazing that it's not being fixed.  https://issues.apache.org/jira/browse/MNG-7003

**Comment by INPUTsys-Chris** [comment]
@dbwiddis Thank you for your research. And thank you for putting so much work in Oshi in general.

The "sample project" is in the linked Vaadin issue under "Minimal reproducible example". Sorry for not being accurate.

I think I'll find a way around that behaviour. Sorry for reporting it as a Oshi "bug" while it's a maven one.

**Comment by dbwiddis** [comment]
No worries, even though it's a maven bug we should work around it to avoid issues like this.   The latest maven JIRA linked above (7003) mentions a flatten dependency plugin that's also a workaround.

I'll get out a new version of OSHI in the next day or two to resolve this... I'm thinking of just not using dependency management at all since outside of our own build it seems ineffective.

**Comment by hazendaz** [comment]
Nothing recent.  It's how it works.  Dependency management is used by many such as spingboot.  It works fine.  Order in your own project pom matters.  Even if we use dependency management, moving it won't solve the issue.  Use dependency management of oshi in your pom with type pom and scope import will lock it.  Again order matters there too. With so many using dependency management now it's a fine balance.

If I recall this is further documented on oshi for same reason.  Or at least it was when I documented same at waffle.  At any rate it's always the above to resolve.  There is a lot of self control there aside from any actual maven bugs otherwise pointed out.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by INPUTsys-Chris** [comment]
> It's how it works. Dependency management is used by many such as spingboot. It works fine.

There may be a reason why Oshi itself says
"Users of the Spring Boot Starter Parent should include this property in their POM"
here
https://github.com/oshi/oshi/blob/master/pom.xml

**Comment by dbwiddis** [comment]
Thanks, @hazendaz 
> Nothing recent.  

Yeah, I see the unexpected behavior documented since 2015...

> Use dependency management of oshi in your pom with type pom and scope import will lock it.

Should OSHI be using scope import for JNA?  Currently we just have this:
```xml
<dependencyManagement>
    <dependencies>
        <!-- Due to critical nature of OSHI jna usage, set to dependency management to help influence usage -->
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna</artifactId>
            <version>${jna.version}</version>
        </dependency>
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna-platform</artifactId>
            <version>${jna.version}</version>
        </dependency>
    </dependencies>
</dependencyManagement>
```
Should we change it to this?
```xml
<dependencyManagement>
    <dependencies>
        <!-- Due to critical nature of OSHI jna usage, set to dependency management to help influence usage -->
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna</artifactId>
            <version>${jna.version}</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna-platform</artifactId>
            <version>${jna.version}</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
```
This note on another vaadin project is also suspicious:  https://github.com/vaadin/osgi/blob/e240aaaa3728d6d1d5e08f7427458c3c87fc2c3b/test-containers/felix-jetty/pom.xml#L170-L175
```xml
<!-- oshi-core dependency that doesn't get added for some reason -->
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna</artifactId>
    <version>5.11.0</version>
</dependency>
```
"For some reason" seems to indicate we are doing something wrong with specifying the JNA dependency that I don't understand.

TLDR: I think an appropriate fix is to explicitly add the `${jna.version}` version for JNA in the `<dependencies>` section in addition to the `<dependencyManagement>` section.  Is there a good reason not to do this?

@INPUTsys-Chris, re:
> "Users of the Spring Boot Starter Parent should include this property in their POM"

Spring Boot's parent itself has a jna.version property and there was no way to put OSHI first there. There may be something similar going on with vaadin and JNA dependency management.

**Comment by hazendaz** [comment]
Pretty sure no change needed in oshi.  I can take a deeper look at this ticket in a couple days.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by dbwiddis** [comment]
> Pretty sure no change needed in oshi. I can take a deeper look at this ticket in a couple days.

Thanks. I actually started trying a few ideas but ended up leaving it as is:
  - Removing the whole `dependencyManagement` section.  Works for this use case and some others but prevents the "optimal" solution of "import" from another project.
  - Leaving dependency management but adding in versions. Got warnings about overrides.  Which is fine, but I always hate overrides and no obvious "annotations" in POM files to say "yeah I know"
  - Moving dependency management to the parent POM.  Same as previous but apparently more "canonical" but given -core and -core-java11 use different dependencies didn't gain anything.

At the end of the day left as is for now.  At a bare minimum we can add to docs/FAQ the best option here.  Appreciate you looking into this whenever you get the time.

**Comment by INPUTsys-Chris** [comment]
Thank you for your work. We now handle JNA dependency on our own.
Any improvement would be good, but from our side you could close the ticket
